keps/sig-node/20200310-ensure-secret-pulled-images.md

```diff
@@ -0,0 +1,190 @@
+[#18787](https://github.com/kubernetes/kubernetes/issues/18787) has been around
+for a while.
+
+To secure images one currently needs to inject `AllwaysPullImages` into pod
+specs via an admission plugin. As @liggitt [notes](https://github.com/kubernetes/kubernetes/issues/18787#issuecomment-532280931)
+the `pull` does not re-pull already-pulled layers of the image, but simply
+resolves/verifies the image manifest has not changed in the registry (which
+incidentally requires authenticating to private registries, which enforces the
+image access). That means in the normal case (where the image has not changed
+since the last pull), the request size is O(kb). However, the `pull` does put
+the registry in the critical path of starting a container, since an unavailable
+registry will fail the pull image manifest check (with or without proper
+authentication.)
+
+
+### Goals
+
+Modify the current pullIfNotPresent policy management enforced by `kubelet` to
+ensure the images pulled with a secret by `kublet` since boot. During the
+EnsureImagesExist step `kubelet` will require authentication of present images
+pulled with auth since boot.  
+
+Optimize to only force re-authentication for a pod container image when the
+secret used to pull the container image is not present. IOW if an image is
+pulled with authentication for a first pod, subsequent pods that have the same
+authentication information should not need to re-authenticate.
+
+Images already present at boot or loaded externally to `kubelet` or successfully
+pulled through `kubelet` with no imagePullSecret/authentication required will
+not require authentication.
+
+### Non-Goals
+
+Out of scope for this KEP is an image caching policy that would direct container
+runtimes through the CRI wrt. how they should treat the caching of images on a
+node. Such as store for public use but only if encrypted. Or Store for private
+use un-encrypted...
+
+## Proposal
+
+`kubelet` will keep a list, since boot, of container images that required
+authentication and a list of the authentications that successfully pulled the image.
+
+`kubelet` will ensure any image in the list is always pulled if an authentication
+used is not present, thus enforcing authentication / re-authentication.
+
+
+### User Stories
+wip
+
+### Risks and Mitigations
+
+Image authentications with a registry may expire. To mitigate expirations a
+a timeout could be used to force re-authentication. The timeout could be a
+container runtime feature or a `kubelet` feature. If at the container runtime,
+images would not be present during the EnsureImagesExist step, thus would have
+to be pulled and authenticated if necessary.  
+
+Since images can be pre-loaded, loaded outside the `kubelet` process, and
+garbage collected.. the list of images that required authentication in `kubelet`
+will not be a source of truth for how all images were pulled that are in the
+container runtime cache. To mitigate images can be garbage collected at boot.
```

Comment:
at boot or startup of kubelet?

Reply:
either would work for this purpose.. I know containerd can also perform gc.